File: src/grid_map.cc
#include "endfield_base/grid_map.h"

#include <algorithm>
#include <ranges>
#include <stdexcept>

namespace endfield_base {
GridMap::GridMap(int width, int height) {
    reset(width, height);
}

int GridMap::getWidth() const {
    return width_;
}

int GridMap::getHeight() const {
    return height_;
}

int GridMap::getNextInstanceId() const {
    return nextInstanceId_;
}

void GridMap::setNextInstanceId(int nextInstanceId) {
    nextInstanceId_ = nextInstanceId;
}

const std::vector<FacilityInstance>& GridMap::getFacilities() const {
    return facilities_;
}

FacilityInstance* GridMap::findFacility(int instanceId) {
    const auto it = std::ranges::find_if(
        facilities_,
        [instanceId] (const FacilityInstance& instance) { return instance.instanceId == instanceId; }
    );

    if (it == facilities_.end()) {
        return nullptr;
    }

    return &*it;
}

const FacilityInstance* GridMap::findFacility(int instanceId) const {
    const auto it = std::ranges::find_if(
        facilities_,
        [instanceId] (const FacilityInstance& instance) { return instance.instanceId == instanceId; }
    );

    if (it == facilities_.end()) {
        return nullptr;
    }

    return &*it;
}

const CellOccupancy* GridMap::getCell(int x, int y) const {
    if (x < 0 || y < 0 || x >= width_ || y >= height_) {
        return nullptr;
    }

    return &cells_.at(indexForCell(x, y));
}

auto GridMap::getOccupiedCells(
    const FacilityDefinition& definition,
    const FacilityInstance& instance
) const -> std::vector<GridPoint> {
    const GridSize footprint = rotateFootprint(definition.footprint, instance.rotation);
    std::vector<GridPoint> occupiedCells;
    occupiedCells.reserve(static_cast<std::size_t>(footprint.width * footprint.height));
    for (int offsetY = 0; offsetY < footprint.height; ++offsetY) {
        for (int offsetX = 0; offsetX < footprint.width; ++offsetX) {
            occupiedCells.push_back({instance.position.x + offsetX, instance.position.y + offsetY});
        }
    }
    return occupiedCells;
}

std::vector<int> GridMap::getFacilityIdsAt(const GridPoint& cell) const {
    const CellOccupancy* occupancy = getCell(cell.x, cell.y);
    if (occupancy == nullptr) {
        return {};
    }

    std::vector<int> ids;
    ids.reserve(
        occupancy->attachedInstanceIds.size()
        + static_cast<std::size_t>(occupancy->baseInstanceId.has_value())
        + static_cast<std::size_t>(occupancy->bridgeInstanceId.has_value())
    );
    if (const auto baseInstanceId = occupancy->baseInstanceId) {
        ids.push_back(*baseInstanceId);
    }
    if (const auto bridgeInstanceId = occupancy->bridgeInstanceId) {
        ids.push_back(*bridgeInstanceId);
    }
    ids.insert(ids.end(), occupancy->attachedInstanceIds.begin(), occupancy->attachedInstanceIds.end());
    return ids;
}

auto GridMap::canPlaceFacility(
    const FacilityDefinition& definition,
    const GridPoint& origin,
    Rotation rotation,
    std::string* reason
) const -> bool {
    const GridSize footprint = rotateFootprint(definition.footprint, rotation);
    if (origin.x < 0 || origin.y < 0 || origin.x + footprint.width > width_ || origin.y + footprint.height >
        height_) {
        if (reason != nullptr) {
            *reason = "placement out of bounds";
        }
        return false;
    }

    for (int offsetY = 0; offsetY < footprint.height; ++offsetY) {
        for (int offsetX = 0; offsetX < footprint.width; ++offsetX) {
            const GridPoint currentCell{origin.x + offsetX, origin.y + offsetY};
            const CellOccupancy& occupancy = cells_.at(indexForCell(currentCell.x, currentCell.y));

            if (!definition.occupiesBaseGrid) {
                if (definition.canAttachToStorageLine && !isPerimeterCell(currentCell)) {
                    if (reason != nullptr) {
                        *reason = "storage port must attach to perimeter line";
                    }
                    return false;
                }

                if (definition.canMountOnConveyor && !occupancy.baseInstanceId.has_value()) {
                    if (reason != nullptr) {
                        *reason = "mounted facility requires base carrier";
                    }
                    return false;
                }

                if (!occupancy.attachedInstanceIds.empty()) {
                    if (reason != nullptr) {
                        *reason = "attachment layer already occupied";
                    }
                    return false;
                }

                continue;
            }

            if (supportsBridgeLayer(definition.category)) {
                if (occupancy.bridgeInstanceId.has_value()) {
                    if (reason != nullptr) {
                        *reason = "bridge layer already occupied";
                    }
                    return false;
                }
            } else {
                if (occupancy.baseInstanceId.has_value()) {
                    if (reason != nullptr) {
                        *reason = "base layer already occupied";
                    }
                    return false;
                }
            }
        }
    }

    return true;
}

auto GridMap::placeFacility(
    const FacilityDefinition& definition,
    std::string_view definitionId,
    const GridPoint& origin,
    Rotation rotation,
    std::optional<int> forcedInstanceId,
    std::string* reason
) -> int {
    if (!canPlaceFacility(definition, origin, rotation, reason)) {
        return 0;
    }

    const int instanceId = forcedInstanceId.value_or(nextInstanceId_);
    nextInstanceId_ = std::max(nextInstanceId_, instanceId + 1);
    facilities_.emplace_back(FacilityInstance{
        instanceId,
        std::string(definitionId),
        origin,
        rotation,
    });
    const GridSize footprint = rotateFootprint(definition.footprint, rotation);

    for (int offsetY = 0; offsetY < footprint.height; ++offsetY) {
        for (int offsetX = 0; offsetX < footprint.width; ++offsetX) {
            CellOccupancy& occupancy = cells_.at(indexForCell(origin.x + offsetX, origin.y + offsetY));
            if (!definition.occupiesBaseGrid) {
                occupancy.attachedInstanceIds.push_back(instanceId);
            } else if (supportsBridgeLayer(definition.category)) {
                occupancy.bridgeInstanceId = instanceId;
            } else {
                occupancy.baseInstanceId = instanceId;
            }
        }
    }

    return instanceId;
}

bool GridMap::removeFacility(int instanceId, const FacilityCatalog& catalog) {
    const auto it = std::ranges::find_if(
        facilities_,
        [instanceId] (const FacilityInstance& instance) { return instance.instanceId == instanceId; }
    );
    if (it == facilities_.end()) {
        return false;
    }

    const FacilityDefinition* definition = catalog.findDefinition(it->definitionId);
    if (definition == nullptr) {
        throw std::runtime_error("Unknown definition id while removing facility: " + it->definitionId);
    }

    for (const auto occupiedCells = getOccupiedCells(*definition, *it); const GridPoint& cell : occupiedCells) {
        CellOccupancy& occupancy = cells_.at(indexForCell(cell.x, cell.y));
        if (!definition->occupiesBaseGrid) {
            std::erase(
                occupancy.attachedInstanceIds,
                instanceId
            );
        } else if (supportsBridgeLayer(definition->category)) {
            occupancy.bridgeInstanceId.reset();
        } else {
            occupancy.baseInstanceId.reset();
        }
    }

    facilities_.erase(it);
    return true;
}

void GridMap::reset(int width, int height) {
    width_ = width;
    height_ = height;
    nextInstanceId_ = 1;
    facilities_.clear();
    cells_.assign(static_cast<std::size_t>(width_ * height_), {});
}

std::size_t GridMap::indexForCell(int x, int y) const {
    return static_cast<std::size_t>(y * width_ + x);
}

bool GridMap::isInsideBounds(const GridPoint& cell) const {
    return cell.x >= 0 && cell.y >= 0 && cell.x < width_ && cell.y < height_;
}

bool GridMap::isPerimeterCell(const GridPoint& cell) const {
    if (!isInsideBounds(cell)) {
        return false;
    }

    return cell.x == 0 || cell.y == 0 || cell.x == width_ - 1 || cell.y == height_ - 1;
}

void GridMap::rebuildOccupancy(const FacilityCatalog& catalog) {
    cells_.assign(static_cast<std::size_t>(width_ * height_), {});
    for (const FacilityInstance& instance : facilities_) {
        const FacilityDefinition* definition = catalog.findDefinition(instance.definitionId);
        if (definition == nullptr) {
            throw std::runtime_error("Unknown definition id in map: " + instance.definitionId);
        }

        for (const auto occupiedCells = getOccupiedCells(*definition, instance);
             const GridPoint& cell : occupiedCells) {
            CellOccupancy& occupancy = cells_.at(indexForCell(cell.x, cell.y));
            if (!definition->occupiesBaseGrid) {
                occupancy.attachedInstanceIds.push_back(instance.instanceId);
            } else if (supportsBridgeLayer(definition->category)) {
                occupancy.bridgeInstanceId = instance.instanceId;
            } else {
                occupancy.baseInstanceId = instance.instanceId;
            }
        }
    }
}
}  // namespace endfield_base

File: src/EndfieldBase.Core/Grid/GridMap.cs
using EndfieldBase.Core.Models;

namespace EndfieldBase.Core.Grid;

public sealed class CellOccupancy
{
    public int? BaseInstanceId { get; set; }
    public int? BridgeInstanceId { get; set; }
    public List<int> AttachedInstanceIds { get; } = [];
}

public sealed class GridMap
{
    private int _width;
    private int _height;
    private int _nextInstanceId = 1;
    private readonly List<FacilityInstance> _facilities = [];
    private List<CellOccupancy> _cells = [];

    public GridMap() {}

    public GridMap(int width, int height) {
        Reset(width, height);
    }

    public int GetWidth() {
        return _width;
    }

    public int GetHeight() {
        return _height;
    }

    public int GetNextInstanceId() {
        return _nextInstanceId;
    }

    public void SetNextInstanceId(int nextInstanceId) {
        _nextInstanceId = nextInstanceId;
    }

    public IReadOnlyList<FacilityInstance> GetFacilities() {
        return _facilities;
    }

    public FacilityInstance? FindFacility(int instanceId) {
        return _facilities.FirstOrDefault(instance => instance.InstanceId == instanceId);
    }

    public CellOccupancy? GetCell(int x, int y) {
        return x < 0 || y < 0 || x >= _width || y >= _height ? null : _cells[IndexForCell(x, y)];
    }

    public List<GridPoint> GetOccupiedCells(FacilityDefinition definition, FacilityInstance instance) {
        var footprint = FacilityRules.RotateFootprint(definition.Footprint, instance.Rotation);
        List<GridPoint> occupiedCells = new(footprint.Width * footprint.Height);
        for (var offsetY = 0; offsetY < footprint.Height; offsetY++)
            for (var offsetX = 0; offsetX < footprint.Width; offsetX++)
                occupiedCells.Add(new GridPoint(instance.Position.X + offsetX, instance.Position.Y + offsetY));

        return occupiedCells;
    }

    public List<int> GetFacilityIdsAt(GridPoint cell) {
        var occupancy = GetCell(cell.X, cell.Y);
        if (occupancy is null) return [];

        List<int> ids = new(occupancy.AttachedInstanceIds.Count + (occupancy.BaseInstanceId.HasValue ? 1 : 0) +
                            (occupancy.BridgeInstanceId.HasValue ? 1 : 0));
        if (occupancy.BaseInstanceId.HasValue) ids.Add(occupancy.BaseInstanceId.Value);

        if (occupancy.BridgeInstanceId.HasValue) ids.Add(occupancy.BridgeInstanceId.Value);

        ids.AddRange(occupancy.AttachedInstanceIds);
        return ids;
    }

    public bool CanPlaceFacility(
        FacilityDefinition definition,
        GridPoint origin,
        Rotation rotation,
        out string? reason) {
        var footprint = FacilityRules.RotateFootprint(definition.Footprint, rotation);
        if (origin.X < 0 || origin.Y < 0 || origin.X + footprint.Width > _width ||
            origin.Y + footprint.Height > _height) {
            reason = "placement out of bounds";
            return false;
        }

        for (var offsetY = 0; offsetY < footprint.Height; offsetY++)
            for (var offsetX = 0; offsetX < footprint.Width; offsetX++) {
                GridPoint currentCell = new(origin.X + offsetX, origin.Y + offsetY);
                var occupancy = _cells[IndexForCell(currentCell.X, currentCell.Y)];

                if (!definition.OccupiesBaseGrid) {
                    if (definition.CanAttachToStorageLine && !IsPerimeterCell(currentCell)) {
                        reason = "storage port must attach to perimeter line";
                        return false;
                    }

                    if (definition.CanMountOnConveyor && !occupancy.BaseInstanceId.HasValue) {
                        reason = "mounted facility requires base carrier";
                        return false;
                    }

                    if (occupancy.AttachedInstanceIds.Count > 0) {
                        reason = "attachment layer already occupied";
                        return false;
                    }

                    continue;
                }

                if (FacilityRules.SupportsBridgeLayer(definition.Category)) {
                    if (occupancy.BridgeInstanceId.HasValue) {
                        reason = "bridge layer already occupied";
                        return false;
                    }
                }
                else if (occupancy.BaseInstanceId.HasValue) {
                    reason = "base layer already occupied";
                    return false;
                }
            }

        reason = null;
        return true;
    }

    public int PlaceFacility(
        FacilityDefinition definition,
        string definitionId,
        GridPoint origin,
        Rotation rotation,
        out string? reason,
        int? forcedInstanceId = null) {
        if (!CanPlaceFacility(definition, origin, rotation, out reason)) return 0;

        var instanceId = forcedInstanceId ?? _nextInstanceId;
        _nextInstanceId = Math.Max(_nextInstanceId, instanceId + 1);
        FacilityInstance instance = new()
            {
                InstanceId = instanceId,
                DefinitionId = definitionId,
                Position = origin,
                Rotation = rotation
            };
        _facilities.Add(instance);

        var footprint = FacilityRules.RotateFootprint(definition.Footprint, rotation);
        for (var offsetY = 0; offsetY < footprint.Height; offsetY++)
            for (var offsetX = 0; offsetX < footprint.Width; offsetX++) {
                var occupancy = _cells[IndexForCell(origin.X + offsetX, origin.Y + offsetY)];
                if (!definition.OccupiesBaseGrid)
                    occupancy.AttachedInstanceIds.Add(instanceId);
                else if (FacilityRules.SupportsBridgeLayer(definition.Category))
                    occupancy.BridgeInstanceId = instanceId;
                else
                    occupancy.BaseInstanceId = instanceId;
            }

        return instanceId;
    }

    public bool RemoveFacility(int instanceId, FacilityCatalog catalog) {
        var instance = FindFacility(instanceId);
        if (instance is null) return false;

        var definition = catalog.FindDefinition(instance.DefinitionId)
                         ?? throw new InvalidOperationException(
                             $"Unknown definition id while removing facility: {instance.DefinitionId}");

        foreach (var cell in GetOccupiedCells(definition, instance)) {
            var occupancy = _cells[IndexForCell(cell.X, cell.Y)];
            if (!definition.OccupiesBaseGrid)
                occupancy.AttachedInstanceIds.Remove(instanceId);
            else if (FacilityRules.SupportsBridgeLayer(definition.Category))
                occupancy.BridgeInstanceId = null;
            else
                occupancy.BaseInstanceId = null;
        }

        _facilities.Remove(instance);
        return true;
    }

    public void Reset(int width, int height) {
        _width = width;
        _height = height;
        _nextInstanceId = 1;
        _facilities.Clear();
        _cells = Enumerable.Range(0, width * height).Select(_ => new CellOccupancy()).ToList();
    }

    public void RebuildOccupancy(FacilityCatalog catalog) {
        _cells = Enumerable.Range(0, _width * _height).Select(_ => new CellOccupancy()).ToList();
        foreach (var instance in _facilities) {
            var definition = catalog.FindDefinition(instance.DefinitionId)
                             ?? throw new InvalidOperationException(
                                 $"Unknown definition id in map: {instance.DefinitionId}");

            foreach (var cell in GetOccupiedCells(definition, instance)) {
                var occupancy = _cells[IndexForCell(cell.X, cell.Y)];
                if (!definition.OccupiesBaseGrid)
                    occupancy.AttachedInstanceIds.Add(instance.InstanceId);
                else if (FacilityRules.SupportsBridgeLayer(definition.Category))
                    occupancy.BridgeInstanceId = instance.InstanceId;
                else
                    occupancy.BaseInstanceId = instance.InstanceId;
            }
        }
    }

    private int IndexForCell(int x, int y) {
        return y * _width + x;
    }

    private bool IsInsideBounds(GridPoint cell) {
        return cell.X >= 0 && cell.Y >= 0 && cell.X < _width && cell.Y < _height;
    }

    private bool IsPerimeterCell(GridPoint cell) {
        return IsInsideBounds(cell)
               && (cell.X == 0 || cell.Y == 0 || cell.X == _width - 1 || cell.Y == _height - 1);
    }
}